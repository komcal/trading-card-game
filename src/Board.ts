import { Player } from './Player'

interface BoardArgs {
    player1: Player
    player2: Player
}

class Board {
    player1: Player
    player2: Player
    start: boolean
    private currentPlayerKey: 1 | 2
    maxHandSize: number
    constructor(args: BoardArgs) {
        this.player1 = args.player1
        this.player2 = args.player2
        this.start = false
        this.maxHandSize = 5
    }
    public startGame() {
        if (!this.player1 || !this.player2) {
            throw new Error('cannot start game. waiting for other player')
        }
        this.start = true
        this.currentPlayerKey = Math.floor(Math.random() * 2) + 1 as 1 | 2
    }
    get currentPlayer() {
        if (!this.currentPlayerKey) {
            return null
        }
        if (this.currentPlayerKey === 1) {
            return this.player1
        }
        return this.player2
    }
    public drawPhase(amount: number) {
        try {
            const cards = this.currentPlayer.drawCard(amount)
            if (this.currentPlayer.handCardCount < this.maxHandSize) {
                this.currentPlayer.addCardToHand(cards)
            }
        } catch(e) {
            this.currentPlayer.bleedingOut()
        }
    }
}

export {
    Board
}