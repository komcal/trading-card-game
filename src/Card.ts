import { Player } from './Player'
interface CardArgs {
  mana: number
}
class Card {
  mana: number
  constructor(args: CardArgs) {
    this.mana = args.mana
  }
  public action(enemyPlayer: Player) {
    enemyPlayer.health -= this.mana
  }
}

export {
  Card
}