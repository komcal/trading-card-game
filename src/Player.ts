import { Card } from "./Card";

interface PlayerArgs {
    name: string
    cards: Card[]
}
class Player {
    maxMana: number
    currentMana: number
    handcards: Card[]
    cards: Card[]
    health: number
    name: string
    mana: number
    constructor(args: PlayerArgs) {
        this.maxMana = 0
        this.currentMana = 0
        this.mana = 0
        this.cards = args.cards
        this.health = 30
        this.name = args.name
        this.handcards = []
    }
    public bleedingOut() {
        this.health--
    }
    public drawCard(amount: number) {
        const len = this.cards.length
        if (len === 0) {
            throw new Error('Cannot draw card')
        }
        const cards = this.cards.splice(0, amount)
        return cards
    }
    public useCard(number: number) {
        const card = this.handcards[number]
        if (!card || card.mana > this.currentMana) {
            throw new Error(`cannot use this card number ${number}`)
        }
        this.handcards.splice(number, 1)
        this.currentMana -= card.mana
        return card
    }
    public addCardToHand(cards: Card[]) {
        this.handcards.push(...cards)
    }
    get handCardCount() {
        return this.handcards.length
    }
}

export {
    Player
}