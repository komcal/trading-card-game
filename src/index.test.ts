import { Player } from './Player'
import { Board } from './Board'
import { Card } from './Card'
describe('Initial', () => {
    it('should random start player when start game', () => {
       const player1 = new Player({
           name: 'a',
           cards: []
       })
       const player2 = new Player({
           name: 'b',
           cards: []
       })
       const board = new Board({
           player1,
           player2
       })
       expect(board.currentPlayer).toBeNull()
       board.startGame(1)
       expect(board.currentPlayer).toBeDefined()
    })
})
describe('Draw phase', () => {
    it('should draw 3 cards', () => {
       const player1 = new Player({
           name: 'a',
           cards: [
               new Card({ mana: 0 }),
               new Card({ mana: 0 }),
               new Card({ mana: 0 }),
           ]
       })
       const player2 = new Player({
           name: 'b',
           cards: []
       })
       const board = new Board({
           player1,
           player2
       })
       board.startGame(1)
       board.drawPhase(3)
       expect(board.currentPlayer.cards.length).toEqual(0)
       expect(board.currentPlayer.handCardCount).toEqual(3)
    })
    it('should not draw card when card on hand equal maxHandSize', () => {
       const player1 = new Player({
           name: 'a',
           cards: [
               new Card({ mana: 0 }),
               new Card({ mana: 0 }),
               new Card({ mana: 0 }),
               new Card({ mana: 0 }),
               new Card({ mana: 0 }),
               new Card({ mana: 0 }),
               new Card({ mana: 0 }),
            ]
       })
       const player2 = new Player({
           name: 'b',
           cards: []
       })
       const board = new Board({
           player1,
           player2
       })
       board.startGame(1)
       board.drawPhase(3)
       board.drawPhase(1)
       board.drawPhase(1)
       board.drawPhase(1)
       board.drawPhase(1)
       expect(board.currentPlayer.cards.length).toEqual(0)
       expect(board.currentPlayer.handCardCount).toEqual(board.maxHandSize)
    })
    it('should bleeding out when no card to draw', () => {
       const player1 = new Player({
           name: 'a',
           cards: []
       })
       const player2 = new Player({
           name: 'b',
           cards: []
       })
       const board = new Board({
           player1,
           player2
       })
       board.startGame(1)
       board.drawPhase(1)
       expect(board.currentPlayer.cards.length).toEqual(0)
       expect(board.currentPlayer.handCardCount).toEqual(0)
       expect(board.currentPlayer.health).toEqual(29)
    })
})
describe('snapshots test', () => {
    const player1 = new Player({
        name: 'a',
        cards: [
            new Card({ mana: 3 }),
            new Card({ mana: 0 }),
            new Card({ mana: 1 }),
            new Card({ mana: 5 }),
            new Card({ mana: 2 }),
            new Card({ mana: 3 }),
            new Card({ mana: 6 }),
            new Card({ mana: 0 }),
            new Card({ mana: 3 }),
            new Card({ mana: 4 }),
            new Card({ mana: 3 }),
            new Card({ mana: 1 }),
            new Card({ mana: 2 }),
            new Card({ mana: 4 }),
            new Card({ mana: 8 }),
            new Card({ mana: 2 }),
            new Card({ mana: 5 }),
            new Card({ mana: 6 }),
            new Card({ mana: 4 }),
            new Card({ mana: 7 }),
        ]
    })
    const player2 = new Player({
        name: 'b',
        cards: [
            new Card({ mana: 1 }),
            new Card({ mana: 0 }),
            new Card({ mana: 6 }),
            new Card({ mana: 2 }),
            new Card({ mana: 5 }),
            new Card({ mana: 3 }),
            new Card({ mana: 4 }),
            new Card({ mana: 3 }),
            new Card({ mana: 2 }),
            new Card({ mana: 6 }),
            new Card({ mana: 3 }),
            new Card({ mana: 4 }),
            new Card({ mana: 0 }),
            new Card({ mana: 1 }),
            new Card({ mana: 4 }),
            new Card({ mana: 3 }),
            new Card({ mana: 8 }),
            new Card({ mana: 2 }),
            new Card({ mana: 7 }),
            new Card({ mana: 5 }),
        ]
    })
    const board = new Board({
        player1,
        player2
    })
    const snap: any[] = []
    //1
    board.startGame(1)
    board.drawPhase(3)
    board.initialTurnPhase()
    board.actions([2, 1])
    snap.push({
        player1: player1.toJson(),
        player2: player2.toJson(),
        isEnd: board.isGameEnd()
    })
    board.endPhase()
    //2
    board.drawPhase(3)
    board.initialTurnPhase()
    board.actions([0,0])
    snap.push({
        player1: player1.toJson(),
        player2: player2.toJson(),
        isEnd: board.isGameEnd()
    })
    board.endPhase()
    //1
    board.drawPhase(1)
    board.initialTurnPhase()
    snap.push({
        player1: player1.toJson(),
        player2: player2.toJson(),
        isEnd: board.isGameEnd()
    })
    board.endPhase()
    //2
    board.drawPhase(1)
    board.initialTurnPhase()
    board.actions([1])
    snap.push({
        player1: player1.toJson(),
        player2: player2.toJson(),
        isEnd: board.isGameEnd()
    })
    board.endPhase()
    //1
    board.drawPhase(1)
    board.initialTurnPhase()
    board.actions([0])
    snap.push({
        player1: player1.toJson(),
        player2: player2.toJson(),
        isEnd: board.isGameEnd()
    })
    board.endPhase()
    //2
    board.drawPhase(1)
    board.initialTurnPhase()
    snap.push({
        player1: player1.toJson(),
        player2: player2.toJson(),
        isEnd: board.isGameEnd()
    })
    board.endPhase()
    //1
    board.drawPhase(1)
    board.initialTurnPhase()
    board.actions([2])
    snap.push({
        player1: player1.toJson(),
        player2: player2.toJson(),
        isEnd: board.isGameEnd()
    })
    board.endPhase()
    //2
    board.drawPhase(1)
    board.initialTurnPhase()
    board.actions([2])
    snap.push({
        player1: player1.toJson(),
        player2: player2.toJson(),
        isEnd: board.isGameEnd()
    })
    board.endPhase()
    //1
    board.drawPhase(1)
    board.initialTurnPhase()
    board.actions([0])
    snap.push({
        player1: player1.toJson(),
        player2: player2.toJson(),
        isEnd: board.isGameEnd()
    })
    board.endPhase()
    //2
    board.drawPhase(1)
    board.initialTurnPhase()
    board.actions([1])
    snap.push({
        player1: player1.toJson(),
        player2: player2.toJson(),
        isEnd: board.isGameEnd()
    })
    board.endPhase()
    //1
    board.drawPhase(1)
    board.initialTurnPhase()
    board.actions([1])
    snap.push({
        player1: player1.toJson(),
        player2: player2.toJson(),
        isEnd: board.isGameEnd()
    })
    board.endPhase()
    //2
    board.drawPhase(1)
    board.initialTurnPhase()
    board.actions([0])
    snap.push({
        player1: player1.toJson(),
        player2: player2.toJson(),
        isEnd: board.isGameEnd()
    })
    board.endPhase()
    //1
    board.drawPhase(1)
    board.initialTurnPhase()
    board.actions([0,0,0])
    snap.push({
        player1: player1.toJson(),
        player2: player2.toJson(),
        isEnd: board.isGameEnd()
    })
    board.endPhase()
    //2
    board.drawPhase(1)
    board.initialTurnPhase()
    board.actions([0,0])
    snap.push({
        player1: player1.toJson(),
        player2: player2.toJson(),
        isEnd: board.isGameEnd()
    })
    board.endPhase()
    //1
    board.drawPhase(1)
    board.initialTurnPhase()
    board.actions([0])
    snap.push({
        player1: player1.toJson(),
        player2: player2.toJson(),
        isEnd: board.isGameEnd()
    })
    board.endPhase()
    //2
    board.drawPhase(1)
    board.initialTurnPhase()
    board.actions([0,0])
    snap.push({
        player1: player1.toJson(),
        player2: player2.toJson(),
        isEnd: board.isGameEnd()
    })
    board.endPhase()
    expect(snap).toMatchSnapshot()
})