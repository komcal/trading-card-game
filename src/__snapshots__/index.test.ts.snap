// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Array [
  Object {
    "isEnd": false,
    "player1": Object {
      "handcards": Array [],
      "health": 30,
      "mana": 1,
      "name": "a",
    },
    "player2": Object {
      "handcards": Array [],
      "health": 29,
      "mana": 0,
      "name": "b",
    },
  },
  Object {
    "isEnd": false,
    "player1": Object {
      "handcards": Array [],
      "health": 29,
      "mana": 1,
      "name": "a",
    },
    "player2": Object {
      "handcards": Array [],
      "health": 29,
      "mana": 1,
      "name": "b",
    },
  },
  Object {
    "isEnd": false,
    "player1": Object {
      "handcards": Array [],
      "health": 29,
      "mana": 2,
      "name": "a",
    },
    "player2": Object {
      "handcards": Array [],
      "health": 29,
      "mana": 1,
      "name": "b",
    },
  },
  Object {
    "isEnd": false,
    "player1": Object {
      "handcards": Array [],
      "health": 27,
      "mana": 2,
      "name": "a",
    },
    "player2": Object {
      "handcards": Array [],
      "health": 29,
      "mana": 2,
      "name": "b",
    },
  },
  Object {
    "isEnd": false,
    "player1": Object {
      "handcards": Array [],
      "health": 27,
      "mana": 3,
      "name": "a",
    },
    "player2": Object {
      "handcards": Array [],
      "health": 26,
      "mana": 2,
      "name": "b",
    },
  },
  Object {
    "isEnd": false,
    "player1": Object {
      "handcards": Array [],
      "health": 27,
      "mana": 3,
      "name": "a",
    },
    "player2": Object {
      "handcards": Array [],
      "health": 26,
      "mana": 3,
      "name": "b",
    },
  },
  Object {
    "isEnd": false,
    "player1": Object {
      "handcards": Array [],
      "health": 27,
      "mana": 4,
      "name": "a",
    },
    "player2": Object {
      "handcards": Array [],
      "health": 23,
      "mana": 3,
      "name": "b",
    },
  },
  Object {
    "isEnd": false,
    "player1": Object {
      "handcards": Array [],
      "health": 24,
      "mana": 4,
      "name": "a",
    },
    "player2": Object {
      "handcards": Array [],
      "health": 23,
      "mana": 4,
      "name": "b",
    },
  },
  Object {
    "isEnd": false,
    "player1": Object {
      "handcards": Array [],
      "health": 24,
      "mana": 5,
      "name": "a",
    },
    "player2": Object {
      "handcards": Array [],
      "health": 18,
      "mana": 4,
      "name": "b",
    },
  },
  Object {
    "isEnd": false,
    "player1": Object {
      "handcards": Array [],
      "health": 19,
      "mana": 5,
      "name": "a",
    },
    "player2": Object {
      "handcards": Array [],
      "health": 18,
      "mana": 5,
      "name": "b",
    },
  },
  Object {
    "isEnd": false,
    "player1": Object {
      "handcards": Array [],
      "health": 19,
      "mana": 6,
      "name": "a",
    },
    "player2": Object {
      "handcards": Array [],
      "health": 12,
      "mana": 5,
      "name": "b",
    },
  },
  Object {
    "isEnd": false,
    "player1": Object {
      "handcards": Array [],
      "health": 13,
      "mana": 6,
      "name": "a",
    },
    "player2": Object {
      "handcards": Array [],
      "health": 12,
      "mana": 6,
      "name": "b",
    },
  },
  Object {
    "isEnd": false,
    "player1": Object {
      "handcards": Array [],
      "health": 13,
      "mana": 7,
      "name": "a",
    },
    "player2": Object {
      "handcards": Array [],
      "health": 7,
      "mana": 6,
      "name": "b",
    },
  },
  Object {
    "isEnd": false,
    "player1": Object {
      "handcards": Array [],
      "health": 6,
      "mana": 7,
      "name": "a",
    },
    "player2": Object {
      "handcards": Array [],
      "health": 7,
      "mana": 7,
      "name": "b",
    },
  },
  Object {
    "isEnd": false,
    "player1": Object {
      "handcards": Array [],
      "health": 6,
      "mana": 8,
      "name": "a",
    },
    "player2": Object {
      "handcards": Array [],
      "health": 3,
      "mana": 7,
      "name": "b",
    },
  },
  Object {
    "isEnd": true,
    "player1": Object {
      "handcards": Array [],
      "health": -2,
      "mana": 8,
      "name": "a",
    },
    "player2": Object {
      "handcards": Array [],
      "health": 3,
      "mana": 8,
      "name": "b",
    },
  },
]
`;
